objects/src/accounts/storage.rs

```diff
@@ -1,47 +1,91 @@
-use super::{Digest, Hasher, Word, ZERO};
+use super::{AccountError, BTreeSet, Word};
+use crypto::merkle::{MerkleStore, NodeIndex, SimpleSmt};
 
 // TYPE ALIASES
 // ================================================================================================
 
-pub type StorageItem = (Word, Word);
+pub type StorageItem = (u8, Word);
 
 // ACCOUNT STORAGE
 // ================================================================================================
 
-/// Account storage is a key-value map where both keys and values are words (4 elements).
-///
-/// Internally, account storage uses a compact Sparse Merkle tree to store the data. To ensure that
-/// the data is randomly distributed across the tree, the index of item in the tree is derived by
-/// hashing the key.
-#[derive(Debug, Clone, Default, PartialEq, Eq)]
+/// Account storage is composed of two components. The first component is a simple sparse Merkle
+/// tree of depth 8 which is index addressable. This provides the user with 256 Word slots. Users
+/// that require additional storage can use the second component which is a `MerkleStore`. This
+/// will allow the user to store any Merkle structures they need.  This is achieved by storing the
+/// root of the Merkle structure as a leaf in the simple sparse merkle tree. When `AccountStorage`
+/// is serialized it will check to see if any of the leafs in the simple sparse Merkle tree are
+/// Merkle roots of other Merkle structures.  If any Merkle roots are found then the Merkle
+/// structures will be persisted in the `AccountStorage` `MerkleStore`.
+#[derive(Debug, Clone, PartialEq, Eq)]
 pub struct AccountStorage {
-    // TODO: add compact SMT as backing storage
+    slots: SimpleSmt,
+    store: MerkleStore,
 }
 
 impl AccountStorage {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+
+    /// Depth of the storage tree.
+    pub const STORAGE_TREE_DEPTH: u8 = 8;
+
     // CONSTRUCTOR
     // --------------------------------------------------------------------------------------------
     /// Returns a new instance of account storage initialized with the provided items.
-    pub fn new(_items: &[StorageItem]) -> Self {
-        Self {}
+    pub fn new(
+        items: Vec<StorageItem>,
+        store: MerkleStore,
+    ) -> Result<AccountStorage, AccountError> {
+        // make sure there are no duplicate items for the same index
+        let mut unique = BTreeSet::new();
+        let unique = items.iter().all(move |item| unique.insert(item.0));
+        if !unique {
+            return Err(AccountError::DuplicateStorageItems);
+        }
```

Comment:
Note for the future: ideally, we'd make the error a bit more informative - i.e., include which index is duplicated or something similar - but this is fine for now. Let's create an issue for this so that we don't forget.